File: app/src/main/java/com/youcii/mvplearn/activity/LoginActivity.java
package com.youcii.mvplearn.activity;

import android.content.Intent;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.View;
import android.widget.AutoCompleteTextView;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.jakewharton.rxbinding2.view.RxView;
import com.orhanobut.logger.Logger;
import com.youcii.mvplearn.R;
import com.youcii.mvplearn.activity.interfaces.ILoginView;
import com.youcii.mvplearn.base.BasePresenterActivity;
import com.youcii.mvplearn.presenter.LoginPresenter;
import com.youcii.mvplearn.utils.PermissionUtils;
import com.youcii.mvplearn.utils.PhoneUtils;
import com.youcii.mvplearn.utils.ToastUtils;
import com.youcii.mvplearn.utils.ViewUtils;

import org.jetbrains.annotations.NotNull;

import java.util.concurrent.TimeUnit;

import butterknife.Bind;
import butterknife.ButterKnife;
import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;

/**
 * @author Administrator
 */
public class LoginActivity extends BasePresenterActivity<ILoginView, LoginPresenter> implements ILoginView, TextView.OnEditorActionListener {

    @Bind(R.id.qxjd)
    TextView qxjd;
    @Bind(R.id.login_progress)
    ProgressBar loginProgress;
    @Bind(R.id.bt_login)
    Button login;
    @Bind(R.id.et_user)
    AutoCompleteTextView etUser;
    @Bind(R.id.et_password)
    EditText etPassword;

    @Override
    protected void onCreate(Bundle bundle) {
        super.onCreate(bundle);
        setContentView(R.layout.activity_login);
        Logger.i("Login--onCreate");

        ButterKnife.bind(this);

        etUser.setOnEditorActionListener(this);
        etPassword.setOnEditorActionListener(this);

        subscribeLoginObserver(RxView.clicks(login));
    }

    private void subscribeLoginObserver(Observable<Object> observable) {
        Disposable ignore = observable
                .throttleFirst(1, TimeUnit.SECONDS)
                .subscribe((v) -> {
                    ViewUtils.hideInput(this);
                    presenter.login(etUser.getText().toString(), etPassword.getText().toString());
                });
    }

    @NotNull
    @Override
    public LoginPresenter initPresenter() {
        return new LoginPresenter(this);
    }

    @Override
    protected void onResume() {
        super.onResume();
        Logger.i("Login--onResume");

        // 取消焦点
        qxjd.setFocusable(true);
        qxjd.setFocusableInTouchMode(true);
        qxjd.requestFocus();

        turnLogin(PhoneUtils.isOnLine());

        PermissionUtils.examinePermission(this);
    }

    @Override
    public void loginSuccess() {
        turnProgress(false);
        clearPass();
        startActivity();
    }

    @Override
    public void loginFail(String errorInfo) {
        turnProgress(false);
        showToast("登录失败: " + errorInfo);
    }

    @Override
    public void clearPass() {
        etPassword.setText("");
    }

    @Override
    public void turnProgress(boolean onOff) {
        if (onOff) {
            loginProgress.setVisibility(View.VISIBLE);
        } else {
            loginProgress.setVisibility(View.INVISIBLE);
        }
    }

    @Override
    public void turnLogin(boolean onOff) {
        login.setEnabled(onOff);
    }

    @Override
    public void showToast(String content) {
        ToastUtils.showShortSnack(getWindow().getDecorView(), content);
    }

    @Override
    public void startActivity() {
        Intent intent = new Intent(this, MainActivity.class);
        intent.putExtra("userName", etUser.getText().toString());
        startActivity(intent);
    }

    @Override
    public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
        switch (v.getId()) {
            case R.id.et_user:
                etPassword.requestFocus();
                break;
            case R.id.et_password:
                subscribeLoginObserver(Observable.fromArray(1));
                break;
            default:
                break;
        }
        return true;
    }

    @Override
    protected void onStart() {
        super.onStart();
        Logger.i("Login--onStart");
    }

    @Override
    protected void onPause() {
        super.onPause();
        Logger.i("Login--onPause");
    }

    @Override
    protected void onStop() {
        super.onStop();
        Logger.i("Login--onStop");
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Logger.i("Login--onDestroy");

        // A > B
        // Login--onPause
        // Main--onCreate
        // Main--onStart
        // Main--onResume
        // Login--onStop

        // B > A
        // Main--onPause
        // Login--onStart
        // Login--onResume
        // Main--onStop
        // Main--onDestroy

        // Activity会在onResume后显示
    }
}

File: app/src/main/java/com/youcii/mvplearn/activity/LoginActivity.kt
package com.youcii.mvplearn.activity

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.ImageDecoder
import android.graphics.drawable.AnimatedImageDrawable
import android.os.Build
import android.os.Bundle
import android.view.KeyEvent
import android.view.View
import android.widget.TextView

import com.jakewharton.rxbinding2.view.RxView
import com.orhanobut.logger.Logger
import com.youcii.mvplearn.R
import com.youcii.mvplearn.activity.interfaces.ILoginView
import com.youcii.mvplearn.base.BasePresenterActivity
import com.youcii.mvplearn.presenter.LoginPresenter
import com.youcii.mvplearn.utils.PermissionUtils
import com.youcii.mvplearn.utils.PhoneUtils
import com.youcii.mvplearn.utils.ToastUtils
import com.youcii.mvplearn.utils.ViewUtils

import java.util.concurrent.TimeUnit

import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_login.*

/**
 * @author Administrator
 */
class LoginActivity : BasePresenterActivity<ILoginView, LoginPresenter>(), ILoginView, TextView.OnEditorActionListener {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)
        Logger.i("Login--onCreate")

        // 设置透明状态栏
        // if (Build.VERSION.SDK_INT >= 21) {
        //     // SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN 会保留状态栏, SYSTEM_UI_FLAG_FULLSCREEN 会默认不保留然后下拉显示
        //     // 暂未搞懂 View.SYSTEM_UI_FLAG_LAYOUT_STABLE 的作用
        //     getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN | View.SYSTEM_UI_FLAG_LAYOUT_STABLE);
        //     getWindow().setStatusBarColor(Color.TRANSPARENT);
        //
        //     ActionBar actionBar = getSupportActionBar();
        //     if (actionBar != null) {
        //         actionBar.hide();
        //     }
        // }

        et_user!!.setOnEditorActionListener(this)
        et_password!!.setOnEditorActionListener(this)

        showGifDrawable()

        subscribeLoginObserver(RxView.clicks(bt_login!!))
    }

    private var animatedDrawable: AnimatedImageDrawable? = null

    @SuppressLint("CheckResult")
    private fun showGifDrawable() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            Observable.just(0)
                    .observeOn(Schedulers.io())
                    .map {
                        val resource = ImageDecoder.createSource(resources, R.drawable.chicken)
                        ImageDecoder.decodeDrawable(resource)
                    }.observeOn(AndroidSchedulers.mainThread())
                    .subscribe({ drawable ->
                        tv_gif!!.setImageDrawable(drawable)
                        if (drawable is AnimatedImageDrawable) {
                            animatedDrawable = drawable
                            animatedDrawable!!.repeatCount = 5
                            animatedDrawable!!.start()
                        }
                    }, { error -> ToastUtils.showShortToast(error.toString()) })
            tv_gif.setOnClickListener {
                if (animatedDrawable?.isRunning == false) {
                    animatedDrawable?.start()
                }
            }
        }
    }

    @SuppressLint("CheckResult")
    private fun subscribeLoginObserver(observable: Observable<Any>) {
        observable.throttleFirst(1, TimeUnit.SECONDS)
                .subscribe {
                    ViewUtils.hideInput(this)
                    presenter.login(et_user!!.text.toString(), et_password!!.text.toString())
                }
    }

    override fun initPresenter(): LoginPresenter {
        return LoginPresenter(this)
    }

    override fun onResume() {
        super.onResume()
        Logger.i("Login--onResume")

        // 取消焦点
        qxjd!!.isFocusable = true
        qxjd.isFocusableInTouchMode = true
        qxjd.requestFocus()

        turnLogin(PhoneUtils.isOnLine())

        PermissionUtils.examinePermission(this)
    }

    override fun loginSuccess() {
        turnProgress(false)
        clearPass()
        startActivity()
    }

    override fun loginFail(errorInfo: String) {
        turnProgress(false)
        showToast("登录失败: $errorInfo")
    }

    override fun clearPass() {
        et_password!!.setText("")
    }

    override fun turnProgress(onOff: Boolean) {
        if (onOff) {
            login_progress!!.visibility = View.VISIBLE
        } else {
            login_progress!!.visibility = View.INVISIBLE
        }
    }

    override fun turnLogin(onOff: Boolean) {
        bt_login!!.isEnabled = onOff
    }

    override fun showToast(content: String) {
        ToastUtils.showShortSnack(window.decorView, content)
    }

    override fun startActivity() {
        val intent = Intent(this, MainActivity::class.java)
        intent.putExtra("userName", et_user!!.text.toString())
        startActivity(intent)
    }

    override fun onEditorAction(v: TextView, actionId: Int, event: KeyEvent): Boolean {
        when (v.id) {
            R.id.et_user -> et_password!!.requestFocus()
            R.id.et_password -> subscribeLoginObserver(Observable.fromArray(1))
            else -> {
            }
        }
        return true
    }

    override fun onStart() {
        super.onStart()
        Logger.i("Login--onStart")
    }

    override fun onPause() {
        super.onPause()
        Logger.i("Login--onPause")
    }

    override fun onStop() {
        super.onStop()
        Logger.i("Login--onStop")
    }

    override fun onDestroy() {
        super.onDestroy()
        Logger.i("Login--onDestroy")

        // A > B
        // Login--onPause
        // Main--onCreate
        // Main--onStart
        // Main--onResume
        // Login--onStop

        // B > A
        // Main--onPause
        // Login--onStart
        // Login--onResume
        // Main--onStop
        // Main--onDestroy

        // Activity会在onResume后显示
    }
}

